applications/SystemIdentificationApplication/custom_sensors/measurement_residual_p_norm_response_function.cpp

```diff
@@ -0,0 +1,307 @@
+    }
+};
+
+struct SecondDerivativesGradientCalculation {
+    template <class TEntityType>
+    static inline void Calculate(
+        Sensor& rSensor,
+        Vector& rResponseGradient,
+        const Matrix& rResidualGradient,
+        const TEntityType& rEntity,
+        const ProcessInfo& rProcessInfo)
+    {
+        rSensor.CalculateSecondDerivativesGradient(rEntity, rResidualGradient, rResponseGradient, rProcessInfo);
+    }
+};
+
+struct PartialSensitivity
+{
+    template <class TEntityType, class TVariableType>
+    static inline void Calculate(
+        Sensor& rSensor,
+        Vector& rSensitivityGradient,
+        const Matrix& rSensitivityMatrix,
+        TEntityType& rEntity,
+        const TVariableType& rVariable,
+        const ProcessInfo& rProcessInfo)
+    {
+        rSensor.CalculatePartialSensitivity(rEntity, rVariable, rSensitivityMatrix, rSensitivityGradient, rProcessInfo);
+    }
+};
+
+} // namespace MeasurementResidualPNormResponseFunctionUtilities
+
+MeasurementResidualPNormResponseFunction::MeasurementResidualPNormResponseFunction(const double PCoefficient)
+    : mPCoefficient(PCoefficient)
+{
+    mResponseGradientList.resize(ParallelUtilities::GetNumThreads());
+}
+
+void MeasurementResidualPNormResponseFunction::AddSensor(Sensor::Pointer pSensor)
+{
+    KRATOS_TRY
+
+    KRATOS_ERROR_IF_NOT(pSensor->Has(SENSOR_MEASURED_VALUE))
+        << pSensor->GetName() << " does not have the SENSOR_MEASURED_VALUE defined.\n";
+
+    mpSensorsList.push_back(pSensor);
+
+    KRATOS_CATCH("");
+}
+
+void MeasurementResidualPNormResponseFunction::Clear()
+{
+    mpSensorsList.clear();
+}
+
+std::vector<Sensor::Pointer>& MeasurementResidualPNormResponseFunction::GetSensorsList()
+{
+    return mpSensorsList;
+}
+
+void MeasurementResidualPNormResponseFunction::Initialize()
+{
+    for (auto& p_sensor : mpSensorsList) {
+        p_sensor->Initialize();
+    }
+}
+
+void MeasurementResidualPNormResponseFunction::InitializeSolutionStep()
+{
+    for (auto& p_sensor : mpSensorsList) {
+        p_sensor->InitializeSolutionStep();
+    }
+}
+
+void MeasurementResidualPNormResponseFunction::FinalizeSolutionStep()
+{
+    for (auto& p_sensor : mpSensorsList) {
+        p_sensor->FinalizeSolutionStep();
+    }
+}
+
+void MeasurementResidualPNormResponseFunction::Finalize()
+{
+
+}
+
+double MeasurementResidualPNormResponseFunction::CalculateValue(ModelPart& rModelPart)
+{
+    KRATOS_TRY
+
+    double value = 0.0;
+    for (auto& p_sensor : mpSensorsList) {
+        const double sensor_value = p_sensor->CalculateValue(rModelPart);
+        p_sensor->SetSensorValue(sensor_value);
+        const double current_sensor_error_square  = std::pow(sensor_value - p_sensor->GetValue(SENSOR_MEASURED_VALUE), 2);
+        p_sensor->SetValue(SENSOR_ERROR, current_sensor_error_square);
+        value += std::pow(p_sensor->GetValue(SENSOR_ERROR), mPCoefficient);
+    }
+
+    return std::pow(value, 1 / mPCoefficient);
```

Comment:
Shoudn't this be to be aligned with the `p-norm`?
```suggestion
    return std::pow(value, 1 / (2 * mPCoefficient));
```

Reply:
@sunethwarna thanks for pointing out that I have a mistake in my implementation. Basicaly, we should not put square error in to p, but the error in p. I will go through code again and fix it.